### ISSUE TITLE
Some crates are built twice

### ISSUE THREAD

**Issue by lnicola** [rust-analyzer/rust-analyzer#1925]
This looks like mismatched features, but I'm not sure where:

```
$ cargo run --bin ra_cli --release -- analysis-stats .
  Downloaded once_cell v1.2.0
   Compiling once_cell v1.2.0
   Compiling ra_parser v0.1.0 (crates/ra_parser)
   Compiling ra_prof v0.1.0 (crates/ra_prof)
   Compiling ra_syntax v0.1.0 (crates/ra_syntax)
   Compiling ra_db v0.1.0 (crates/ra_db)
   Compiling ra_mbe v0.1.0 (crates/ra_mbe)
   Compiling ra_fmt v0.1.0 (crates/ra_fmt)
   Compiling ra_project_model v0.1.0 (crates/ra_project_model)
   Compiling ra_hir v0.1.0 (crates/ra_hir)
   Compiling ra_assists v0.1.0 (crates/ra_assists)
   Compiling ra_ide_api v0.1.0 (crates/ra_ide_api)
   Compiling ra_batch v0.1.0 (crates/ra_batch)
   Compiling ra_cli v0.1.0 (crates/ra_cli)
    Finished release [optimized + debuginfo] target(s) in 2m 51s
     Running `target/release/ra_cli analysis-stats .`

$ cargo install-ra                                                     
[snip]

will run: cargo install --path crates/ra_lsp_server --locked --force
  Installing ra_lsp_server v0.1.0 (crates/ra_lsp_server)
    Updating crates.io index
   Compiling matches v0.1.8
   Compiling percent-encoding v2.1.0
   Compiling num-traits v0.2.8
   Compiling rand v0.6.5
   Compiling unicode-normalization v0.1.8
   Compiling threadpool v1.7.1
   Compiling serde_repr v0.1.5
   Compiling lsp-server v0.2.0
   Compiling unicode-bidi v0.3.4
   Compiling proptest v0.9.4
   Compiling num-integer v0.1.41
   Compiling chrono v0.4.9
   Compiling idna v0.2.0
   Compiling flexi_logger v0.14.3
   Compiling url v2.1.0
   Compiling lsp-types v0.61.0
   Compiling ra_text_edit v0.1.0 (crates/ra_text_edit)
   Compiling ra_syntax v0.1.0 (crates/ra_syntax)
   Compiling ra_mbe v0.1.0 (crates/ra_mbe)
   Compiling ra_db v0.1.0 (crates/ra_db)
   Compiling ra_fmt v0.1.0 (crates/ra_fmt)
   Compiling ra_project_model v0.1.0 (crates/ra_project_model)
   Compiling ra_hir v0.1.0 (crates/ra_hir)
   Compiling ra_assists v0.1.0 (crates/ra_assists)
   Compiling ra_ide_api v0.1.0 (crates/ra_ide_api)
   Compiling ra_lsp_server v0.1.0 (crates/ra_lsp_server)
    Finished release [optimized + debuginfo] target(s) in 2m 50s
```

**Comment by matklad** [comment]
It would be cool to pin down the exact crates that cause the problems! And it would be even more cool to establish some meta-procees to automatically find such issues...

**Comment by lnicola** [comment]
Yeah, I've wondered myself about how to debug these. maybe running `cargo -v` would show the enabled features? In this case the first crate that gets rebuilt is `ra_syntax`, so maybe that's where it starts.

Maybe:

1. touch a file in `ra_syntax` or something that's even earlier in the topological order
2. build `ra_cli`
3. build `ra_lsp_server`

I would expect 3 to only build the binary, and none of its dependencies.

**Comment by bjorn3** [comment]
`cargo -Ztimings` shows the enabled features for all crates.

**Comment by bjorn3** [comment]
Does any of its dependencies get a changing feature set?

**Comment by matklad** [comment]
Yeah, `proptest` should be a non-dev dep, as we want to implement Arbitrary. However, we should just get rid of proptest: its way too bloated at the moment: https://github.com/rust-analyzer/rust-analyzer/issues/1720

**Comment by lnicola** [comment]
Probably fixed by #3195... for now.
